direct/src/gui/DirectFrame.py

```diff
@@ -50,166 +52,136 @@ def __init__(self, parent = None, **kw):
         self.defineoptions(kw, optiondefs,
-                           dynamicGroups = DirectFrame.DefDynGroups)
+                           dynamicGroups=DirectFrame.DefDynGroups)
 
         # Initialize superclasses
         DirectGuiWidget.__init__(self, parent)
 
         # Call option initialization functions
         self.initialiseoptions(DirectFrame)
 
-    def destroy(self):
-        DirectGuiWidget.destroy(self)
-
     def clearText(self):
         self['text'] = None
         self.setText()
 
     def setText(self, text=None):
         if text is not None:
-            self['text'] = text
-
-        # Determine if user passed in single string or a sequence
-        if self['text'] == None:
-            textList = (None,) * self['numStates']
-        elif isinstance(self['text'], stringType):
-            # If just passing in a single string, make a tuple out of it
-            textList = (self['text'],) * self['numStates']
+            self["text"] = text
+
+        text = self["text"]
+        component_class = OnscreenText
+        component_kwargs ={
+                "scale": 1,
+                "mayChange": self['textMayChange'],
+                "sort": DGG.TEXT_SORT_INDEX, # this is "30"
+        }
+        
+        if text is None:
+            text_list = (None,) * self['numStates']
         else:
-            # Otherwise, hope that the user has passed in a tuple/list
-            textList = self['text']
-        # Create/destroy components
-        for i in range(self['numStates']):
-            component = 'text' + repr(i)
-            # If fewer items specified than numStates,
-            # just repeat last item
-            try:
-                text = textList[i]
-            except IndexError:
-                text = textList[-1]
-
-            if self.hascomponent(component):
-                if text == None:
-                    # Destroy component
-                    self.destroycomponent(component)
-                else:
-                    self[component + '_text'] = text
+            if self.inputTypeValidation("text", text):#used by button and directmenu
+                text_list = (text,) * self['numStates']
             else:
-                if text == None:
-                    return
-                else:
-                    from .OnscreenText import OnscreenText
-                    self.createcomponent(
-                        component, (), 'text',
-                        OnscreenText,
-                        (), parent = self.stateNodePath[i],
-                        text = text, scale = 1, mayChange = self['textMayChange'],
-                        sort = DGG.TEXT_SORT_INDEX,
-                        )
+                text_list = text
+        
+        self.reinitComponent("text", text_list, component_kwargs, component_class)
 
     def clearGeom(self):
         self['geom'] = None
         self.setGeom()
 
+    def inputTypeValidation(self, name, object_):
```

Comment:
Should this be made private?  Also, try to name methods as a verb, so `validateInputType` would be better